ludwig/datasets/loaders/utils.py

```diff
@@ -0,0 +1,106 @@
+import logging
+import time
+from typing import Any, List, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy
+
+from ludwig.utils.types import DataFrame
+
+
+def _negative_sample_user(interaction_row: np.array, neg_pos_ratio: int, extra_samples: int) -> Tuple[List[int], int]:
+    """Returns a list of negative item indices for given user-item interactions.
+
+    If there are not enough negative items, takes all of them and adds the difference to the extra_samples
+    otherwise, samples with replacement.
+
+    Params:
+        interaction_row: user-item interaction row
+        neg_pos_ratio: number of negative samples per positive sample
+        extra_samples: number of additional samples to add to the negative sample list
+    Returns:
+        Tuple of list of negative item indices and number of extra samples
+    """
+    # Find all items that are not interacted with by the user
+    neg_items = np.where(interaction_row == 0)[1]
+    available_samples = len(neg_items)
+
+    # Randomly sample negative items
+    npos = interaction_row.shape[1] - len(neg_items)
+    samples_required = npos * neg_pos_ratio + extra_samples
+    should_sample = samples_required <= available_samples
+
+    neg_items = np.random.choice(neg_items, samples_required, replace=False) if should_sample else neg_items
+
+    return neg_items.tolist(), max(0, samples_required - available_samples)
+
+
+def negative_sample(
```

Comment:
I think we're exceeding the scope of the datasets API by performing negative sampling here.  The datasets API intent is to return standard benckmark datasets 'as-is' (or as close to the original as we can get).

IMO the sampling implementation belongs in `ludwig.data`, and should be performed as a separate phase after the dataset is loaded (also sampling has hyperparameters `neg_pos_ratio` and `log_pct`).

Is this possible, or would it be too inefficient to merge article and customer features prior to negative sampling?

Reply:
I agree - removed the negative sampling from the loader and moved it to `ludwig.data` in [this PR](https://github.com/ludwig-ai/ludwig/pull/2711)